--- a/llvm/tools/clang/lib/AST/ASTContext.cpp	(revision 157736)
+++ b/llvm/tools/clang/lib/AST/ASTContext.cpp	(working copy)
@@ -6580,7 +6580,7 @@
   if (!getLangOpts().CPlusPlus || FD->hasAttr<GNUInlineAttr>()) {
     // GNU or C99 inline semantics. Determine whether this symbol should be
     // externally visible.
-    if (FD->isInlineDefinitionExternallyVisible())
+//    if (FD->isInlineDefinitionExternallyVisible())
       return External;
 
     // C99 inline semantics, where the symbol is not externally visible.
