--- a/llvm/tools/clang/lib/Sema/SemaDecl.cpp	(revision 157736)
+++ b/llvm/tools/clang/lib/Sema/SemaDecl.cpp	(working copy)
@@ -7369,13 +7369,13 @@
   if (!FD->isLateTemplateParsed())
     CheckForFunctionRedefinition(FD);
 
-  // Builtin functions cannot be defined.
+  /*// Builtin functions cannot be defined.
   if (unsigned BuiltinID = FD->getBuiltinID()) {
     if (!Context.BuiltinInfo.isPredefinedLibFunction(BuiltinID)) {
       Diag(FD->getLocation(), diag::err_builtin_definition) << FD;
       FD->setInvalidDecl();
     }
-  }
+  }*/
 
   // The return type of a function definition must be complete
   // (C99 6.9.1p3, C++ [dcl.fct]p6).
--- a/llvm/tools/clang/lib/AST/ASTContext.cpp	(revision 157736)
+++ b/llvm/tools/clang/lib/AST/ASTContext.cpp	(working copy)
@@ -6548,6 +6548,8 @@
 }
 
 GVALinkage ASTContext::GetGVALinkageForFunction(const FunctionDecl *FD) {
+  return GVA_StrongExternal;
+
   GVALinkage External = GVA_StrongExternal;
 
   Linkage L = FD->getLinkage();
