--- a/dragonegg/src/Types.cpp	2012-03-22 14:48:14.302606248 +0300
+++ b/dragonegg/src/Types.cpp	2012-03-22 14:50:34.153536680 +0300
@@ -699,6 +699,10 @@
     // Compute zext/sext attributes.
     PAttributes |= HandleArgumentExtension(ArgTy);
 
+    // Compute noalias attributes.
+    if (POINTER_TYPE_P(ArgTy) && TYPE_RESTRICT(ArgTy))
+      PAttributes |= Attribute::NoAlias;
+
     if (PAttributes != Attribute::None)
       Attrs.push_back(AttributeWithIndex::get(ArgTys.size(), PAttributes));
   }
@@ -838,11 +842,8 @@
     // inspect it for restrict qualifiers, otherwise try the argument
     // types.
     tree RestrictArgTy = (DeclArgs) ? TREE_TYPE(DeclArgs) : ArgTy;
-    if (TREE_CODE(RestrictArgTy) == POINTER_TYPE ||
-        TREE_CODE(RestrictArgTy) == REFERENCE_TYPE) {
-      if (TYPE_RESTRICT(RestrictArgTy))
-        PAttributes |= Attribute::NoAlias;
-    }
+    if (POINTER_TYPE_P(RestrictArgTy) && TYPE_RESTRICT(RestrictArgTy))
+      PAttributes |= Attribute::NoAlias;
 
 #ifdef LLVM_TARGET_ENABLE_REGPARM
     // Allow the target to mark this as inreg.
