--- a/gcc-4.6/libgomp/Makefile.am	2011-07-10 16:12:28.487269021 +0400
+++ b/gcc-4.6/libgomp/Makefile.am	2011-07-10 16:22:30.200213171 +0400
@@ -39,7 +39,7 @@
 nodist_noinst_HEADERS = libgomp_f.h
 nodist_libsubinclude_HEADERS = omp.h
 if USE_FORTRAN
-nodist_finclude_HEADERS = omp_lib.h omp_lib.f90 omp_lib.mod omp_lib_kinds.mod
+nodist_finclude_HEADERS = omp_lib.h omp_lib.f90 omp_lib.kernelgen.mod omp_lib_kinds.kernelgen.mod
 endif
 
 LTLDFLAGS = $(shell $(SHELL) $(top_srcdir)/../libtool-ldflags $(LDFLAGS))
@@ -47,9 +47,9 @@
 LINK = $(LIBTOOL) --tag CC $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=link \
 	$(CCLD) $(AM_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) $(LTLDFLAGS) -o $@
 
-omp_lib_kinds.mod: omp_lib.mod
+omp_lib_kinds.kernelgen.mod: omp_lib.kernelgen.mod
 	:
-omp_lib.mod: omp_lib.f90
+omp_lib.kernelgen.mod: omp_lib.f90
 	$(FC) $(FCFLAGS) -fsyntax-only omp_lib.f90
 fortran.lo: libgomp_f.h
 fortran.o: libgomp_f.h
--- a/gcc-4.6/libgomp/Makefile.in	2011-09-15 19:34:40.862156116 +0400
+++ b/gcc-4.6/libgomp/Makefile.in	2011-09-15 19:35:52.772256553 +0400
@@ -319,7 +319,7 @@
 
 nodist_noinst_HEADERS = libgomp_f.h
 nodist_libsubinclude_HEADERS = omp.h
-@USE_FORTRAN_TRUE@nodist_finclude_HEADERS = omp_lib.h omp_lib.f90 omp_lib.mod omp_lib_kinds.mod
+@USE_FORTRAN_TRUE@nodist_finclude_HEADERS = omp_lib.h omp_lib.f90 omp_lib.kernelgen.mod omp_lib_kinds.kernelgen.mod
 LTLDFLAGS = $(shell $(SHELL) $(top_srcdir)/../libtool-ldflags $(LDFLAGS))
 LINK = $(LIBTOOL) --tag CC $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=link \
 	$(CCLD) $(AM_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) $(LTLDFLAGS) -o $@
@@ -1079,9 +1079,9 @@
 @LIBGOMP_BUILD_VERSIONED_SHLIB_SUN_TRUE@@LIBGOMP_BUILD_VERSIONED_SHLIB_TRUE@	    sed 's,/\([^/.]*\)\.la,/.libs/\1.a,g'` \
 @LIBGOMP_BUILD_VERSIONED_SHLIB_SUN_TRUE@@LIBGOMP_BUILD_VERSIONED_SHLIB_TRUE@	 > $@ || (rm -f $@ ; exit 1)
 
-omp_lib_kinds.mod: omp_lib.mod
+omp_lib_kinds.kernelgen.mod: omp_lib.kernelgen.mod
 	:
-omp_lib.mod: omp_lib.f90
+omp_lib.kernelgen.mod: omp_lib.f90
 	$(FC) $(FCFLAGS) -fsyntax-only omp_lib.f90
 fortran.lo: libgomp_f.h
 fortran.o: libgomp_f.h
--- a/gcc-4.6/gcc/fortran/module.c	2011-09-15 19:47:07.933199558 +0400
+++ b/gcc-4.6/gcc/fortran/module.c	2011-09-15 19:47:37.723241147 +0400
@@ -76,7 +76,7 @@
 #include "constructor.h"
 #include "cpp.h"
 
-#define MODULE_EXTENSION ".mod"
+#define MODULE_EXTENSION ".kernelgen.mod"
 
 /* Don't put any single quote (') in MOD_VERSION, 
    if yout want it to be recognized.  */
