--- a/gcc-4.5/libgomp/Makefile.am	2011-07-10 16:12:28.487269021 +0400
+++ b/gcc-4.5/libgomp/Makefile.am	2011-07-10 16:22:30.200213171 +0400
@@ -39,7 +39,7 @@
 nodist_noinst_HEADERS = libgomp_f.h
 nodist_libsubinclude_HEADERS = omp.h
 if USE_FORTRAN
-nodist_finclude_HEADERS = omp_lib.h omp_lib.f90 omp_lib.mod omp_lib_kinds.mod
+nodist_finclude_HEADERS = omp_lib.h omp_lib.f90 omp_lib.kernelgen.mod omp_lib_kinds.kernelgen.mod
 endif
 
 LTLDFLAGS = $(shell $(SHELL) $(top_srcdir)/../libtool-ldflags $(LDFLAGS))
@@ -47,9 +47,9 @@
 LINK = $(LIBTOOL) --tag CC $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=link \
 	$(CCLD) $(AM_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) $(LTLDFLAGS) -o $@
 
-omp_lib_kinds.mod: omp_lib.mod
+omp_lib_kinds.kernelgen.mod: omp_lib.kernelgen.mod
 	:
-omp_lib.mod: omp_lib.f90
+omp_lib.kernelgen.mod: omp_lib.f90
 	$(FC) $(FCFLAGS) -fsyntax-only omp_lib.f90
 fortran.lo: libgomp_f.h
 fortran.o: libgomp_f.h
--- a/gcc-4.5/libgomp/Makefile.in	2011-07-10 15:50:44.542459288 +0400
+++ b/gcc-4.5/libgomp/Makefile.in	2011-07-10 16:24:16.699158458 +0400
@@ -351,7 +351,7 @@
 
 nodist_noinst_HEADERS = libgomp_f.h
 nodist_libsubinclude_HEADERS = omp.h
-@USE_FORTRAN_TRUE@nodist_finclude_HEADERS = omp_lib.h omp_lib.f90 omp_lib.mod omp_lib_kinds.mod
+@USE_FORTRAN_TRUE@nodist_finclude_HEADERS = omp_lib.h omp_lib.f90 omp_lib.kernelgen.mod omp_lib_kinds.kernelgen.mod
 LTLDFLAGS = $(shell $(SHELL) $(top_srcdir)/../libtool-ldflags $(LDFLAGS))
 LINK = $(LIBTOOL) --tag CC $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=link \
 	$(CCLD) $(AM_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) $(LTLDFLAGS) -o $@
@@ -1281,9 +1281,9 @@
 
 vpath % $(strip $(search_path))
 
-omp_lib_kinds.mod: omp_lib.mod
+omp_lib_kinds.kernelgen.mod: omp_lib.kernelgen.mod
 	:
-omp_lib.mod: omp_lib.f90
+omp_lib.kernelgen.mod: omp_lib.f90
 	$(FC) $(FCFLAGS) -fsyntax-only omp_lib.f90
 fortran.lo: libgomp_f.h
 fortran.o: libgomp_f.h
--- a/gcc-4.5/gcc/fortran/module.c	2011-07-10 04:02:32.258612162 +0400
+++ b/gcc-4.5/gcc/fortran/module.c	2011-07-10 17:13:10.963201686 +0400
@@ -75,7 +75,7 @@
 #include "md5.h"
 #include "cpp.h"
 
-#define MODULE_EXTENSION ".mod"
+#define MODULE_EXTENSION ".kernelgen.mod"
 
 /* Don't put any single quote (') in MOD_VERSION, 
    if yout want it to be recognized.  */
