--- a/llvm/include/llvm/DerivedTypes.h	(revision 156703)
+++ b/llvm/include/llvm/DerivedTypes.h	(working copy)
@@ -450,6 +450,9 @@
   /// @brief Return the address space of the Pointer type.
   inline unsigned getAddressSpace() const { return getSubclassData(); }
 
+  /// @brief Set the address space of the Pointer type.
+  inline void setAddressSpace(unsigned addrspace) { setSubclassData(addrspace); }
+
   // Implement support type inquiry through isa, cast, and dyn_cast.
   static inline bool classof(const PointerType *) { return true; }
   static inline bool classof(const Type *T) {

