--- a/llvm/lib/Transforms/IPO/PassManagerBuilder.cpp	(revision 192159)
+++ b/llvm/lib/Transforms/IPO/PassManagerBuilder.cpp	(working copy)
@@ -180,7 +180,8 @@
     MPM.add(createScalarReplAggregatesPass(-1, false));
   MPM.add(createEarlyCSEPass());              // Catch trivial redundancies
   MPM.add(createJumpThreadingPass());         // Thread jumps.
-  MPM.add(createCorrelatedValuePropagationPass()); // Propagate conditionals
+  if (OptLevel > 3)
+    MPM.add(createCorrelatedValuePropagationPass()); // Propagate conditionals
   MPM.add(createCFGSimplificationPass());     // Merge & remove BBs
   MPM.add(createInstructionCombiningPass());  // Combine silly seq's
 
@@ -211,7 +212,8 @@
   // opened up by them.
   MPM.add(createInstructionCombiningPass());
   MPM.add(createJumpThreadingPass());         // Thread jumps
-  MPM.add(createCorrelatedValuePropagationPass());
+  if (OptLevel > 3)
+    MPM.add(createCorrelatedValuePropagationPass());
   MPM.add(createDeadStoreEliminationPass());  // Delete dead stores
 
   addExtensionsToPM(EP_ScalarOptimizerLate, MPM);
