--- a/llvm/include/llvm/Support/raw_ostream.h	2012-11-28 19:17:40.798177887 +0100
+++ b/llvm/include/llvm/Support/raw_ostream.h	2012-11-28 19:19:05.058176486 +0100
@@ -310,9 +310,6 @@
 
   uint64_t pos;
 
-  /// write_impl - See raw_ostream::write_impl.
-  virtual void write_impl(const char *Ptr, size_t Size);
-
   /// current_pos - Return the current position within the stream, not
   /// counting the bytes currently in the buffer.
   virtual uint64_t current_pos() const { return pos; }
@@ -324,6 +321,11 @@
   /// been encountered.
   void error_detected() { Error = true; }
 
+protected:
+
+  /// write_impl - See raw_ostream::write_impl.
+  virtual void write_impl(const char *Ptr, size_t Size);
+
 public:
 
   enum {
