--- a/nvopencc/open64/src/be/cg/NVISA/exp_loadstore.cxx
+++ b/nvopencc/open64/src/be/cg/NVISA/exp_loadstore.cxx
@@ -1064,7 +1064,19 @@ Exp_Ldst (
 		base_tn = Gen_Symbol_TN (sym, 0, 0);
 		ofst_tn = Gen_Literal_TN (ofst, 4);
 	} else {
-		FmtAssert(FALSE, ("stack variables not supported"));
+		Set_V_global_mem(variant);
+		if (ST_class(sym) == CLASS_CONST)
+		{
+		  Set_ST_class(sym, CLASS_VAR);
+		  Set_ST_sclass(sym, SCLASS_DGLOBAL);
+		}
+		else
+		  Set_ST_sclass(sym, SCLASS_UGLOBAL);
+		Set_ST_in_global_mem(sym);
+		Set_ST_base(sym, sym);
+		Set_ST_ofst(sym, 0);
+		base_tn = Gen_Symbol_TN (sym, 0, 0);
+		ofst_tn = Gen_Literal_TN (ofst, 4);
 	}
     }
 
--- a/nvopencc/open64/src/be/cg/cgemit.cxx
+++ b/nvopencc/open64/src/be/cg/cgemit.cxx
@@ -8600,11 +8600,14 @@ EMT_Emit_PU ( ST *pu, DST_IDX pu_dst, WN *rwn )
 #endif
 
 #ifdef TARG_NVISA
+    if (!CGEXP_auto_as_static)
+        Process_Bss_Data (CURRENT_SYMTAB);
     CGEMIT_Function_Definition(pu);
     // emit variable decls before the code 
     /* Emit the initialized data associated with this PU. */
     Process_Initos_And_Literals (CURRENT_SYMTAB);
-    Process_Bss_Data (CURRENT_SYMTAB);
+    if (CGEXP_auto_as_static)
+        Process_Bss_Data (CURRENT_SYMTAB);
 #else
     fprintf ( Asm_File, "\n\t%s Program Unit: %s\n", ASM_CMNT, ST_name(pu) );
 #ifdef TARG_SL
