--- a/module_ra_rrtm.f90	2011-08-26 17:57:58.378382392 +0400
+++ b/module_ra_rrtm.f90	2011-08-26 17:52:30.238437697 +0400
@@ -3347,8 +3347,8 @@
                                                         JT, &
                                                        JT1, &
                                                    INDSELF
-      DIMENSION FC00(kts:ktep1),FC01(kts:ktep1),FC10(kts:ktep1),FC11(kts:ktep1)
-      DIMENSION REFPARAM(13)
+      REAL FC00(kts:ktep1),FC01(kts:ktep1),FC10(kts:ktep1),FC11(kts:ktep1)
+      REAL REFPARAM(13)
       DATA REFPARAM/ &
         0.903661, 0.859386, 0.746542, 0.580496, 0.412889, 0.275283, &
         0.162745, 7.63929E-02, 1.82553E-02, 3.72432E-03, &
@@ -3432,7 +3432,7 @@
                                                         JT, &
                                                        JT1, &
                                                    INDSELF
-      DIMENSION H2OREF(59),CO2REF(59), ETAREF(10)
+      REAL H2OREF(59),CO2REF(59), ETAREF(10)
       REAL N2OMULT,N2OREF(59)
       DATA ETAREF/ &
            0.,0.125,0.25,0.375,0.5,0.625,0.75,0.875,0.9875,1.0/
@@ -3931,7 +3931,7 @@
                                                         JT, &
                                                        JT1, &
                                                    INDSELF
-      DIMENSION H2OREF(59),O3REF(59)
+      REAL H2OREF(59),O3REF(59)
       REAL N2OMULT,N2OREF(59)
       DATA H2OREF/ &
            1.87599E-02,1.22233E-02,5.89086E-03,2.76753E-03,1.40651E-03, &
@@ -4049,7 +4049,7 @@
                                                         JT, &
                                                        JT1, &
                                                    INDSELF
-      DIMENSION H2OREF(13),CH4REF(13),ETAREF(11)
+      REAL H2OREF(13),CH4REF(13),ETAREF(11)
       REAL N2OMULT,N2OREF(13)
       DATA N2OREF/ &
            3.20000E-07,3.20000E-07,3.20000E-07,3.20000E-07,3.20000E-07, &
@@ -4603,16 +4603,16 @@
       REAL :: CLRNTTOA,CLRNTSRF
       REAL, PARAMETER :: SECANG=1.66
       REAL, PARAMETER :: WTNUM=0.5
-      DIMENSION BBU(NGPT*(ktep1-kts+1)),BBUTOT(NGPT*(ktep1-kts)),BGLEV(NGPT)
-      DIMENSION PLANKBND(NBANDS),PLNKEMIT(NBANDS)
-      DIMENSION PLVL(NBANDS,0:ktep1),PLAY(NBANDS,kts:ktep1)
-      DIMENSION INDLAY(kts:ktep1),INDLEV(0:ktep1)
-      DIMENSION TLAYFRAC(kts:ktep1),TLEVFRAC(0:ktep1)
-      DIMENSION ABSS(NGPT*(ktep1-kts+1)),ABSCLD(kts:ktep1-1),ATOT(NGPT*(ktep1-kts))
-      DIMENSION ODCLR(NGPT,kts:ktep1-1),ODCLD(kts:ktep1-1),EFCLFRAC(kts:ktep1-1)
-      DIMENSION RADLU(NGPT),RADLD(NGPT)
-      DIMENSION RADCLRU(NGPT),RADCLRD(NGPT)
-      DIMENSION SEMIS(NGPT),RADUEMIT(NGPT)
+      REAL BBU(NGPT*(ktep1-kts+1)),BBUTOT(NGPT*(ktep1-kts)),BGLEV(NGPT)
+      REAL PLANKBND(NBANDS),PLNKEMIT(NBANDS)
+      REAL PLVL(NBANDS,0:ktep1),PLAY(NBANDS,kts:ktep1)
+      INTEGER INDLAY(kts:ktep1),INDLEV(0:ktep1)
+      REAL TLAYFRAC(kts:ktep1),TLEVFRAC(0:ktep1)
+      REAL ABSS(NGPT*(ktep1-kts+1)),ABSCLD(kts:ktep1-1),ATOT(NGPT*(ktep1-kts))
+      REAL ODCLR(NGPT,kts:ktep1-1),ODCLD(kts:ktep1-1),EFCLFRAC(kts:ktep1-1)
+      REAL RADLU(NGPT),RADLD(NGPT)
+      REAL RADCLRU(NGPT),RADCLRD(NGPT)
+      REAL SEMIS(NGPT),RADUEMIT(NGPT)
       INDBOUND = TBOUND - 159.
       TBNDFRAC = TBOUND - INT(TBOUND)
       DO 200 LAY = 0, NLAYERS
