--- a/dragonegg/src/Backend.cpp	2011-07-31 22:40:25.587161725 +0400
+++ b/dragonegg/src/Backend.cpp	2011-07-31 22:29:38.260133658 +0400
@@ -105,6 +105,7 @@
 static bool EnableGCCOptimizations;
 static bool EmitIR;
 static bool SaveGCCOutput;
+bool DisableBuiltins;
 static int LLVMCodeGenOptimizeArg = -1;
 static int LLVMIROptimizeArg = -1;
 
@@ -1952,6 +1953,7 @@
     { "enable-gcc-optzns", &EnableGCCOptimizations },
     { "emit-ir", &EmitIR },
     { "save-gcc-output", &SaveGCCOutput },
+    { "no-builtins", &DisableBuiltins },
     { NULL, NULL } // Terminator.
 };
 
--- a/dragonegg/src/Convert.cpp	2011-07-31 22:40:18.396317069 +0400
+++ b/dragonegg/src/Convert.cpp	2011-07-31 22:32:55.657875740 +0400
@@ -39,6 +39,9 @@
 // System headers
 #include <gmp.h>
 
+// Plugin CLI options
+extern "C" bool DisableBuiltins;
+
 // GCC headers
 extern "C" {
 #include "config.h"
@@ -8515,7 +8518,7 @@
   // Check for a built-in function call.  If we can lower it directly, do so
   // now.
   tree fndecl = gimple_call_fndecl(stmt);
-  if (fndecl && DECL_BUILT_IN(fndecl) &&
+  if (!DisableBuiltins && fndecl && DECL_BUILT_IN(fndecl) &&
       DECL_BUILT_IN_CLASS(fndecl) != BUILT_IN_FRONTEND) {
     Value *Res = 0;
     if (EmitBuiltinCall(stmt, fndecl, DestLoc, Res))
 
